include/boost/thread/detail/thread.hpp

```diff
@@ -641,6 +641,27 @@ namespace boost
             thread_data(other.thread_data)
         {}
 
+        id& operator=(const id& other) BOOST_NOEXCEPT
+        {
+            thread_data = other.thread_data;
+            return *this;
+        }
+
+#if !defined(BOOST_NO_CXX11_RVALUE_REFERENCES)
+        id(id&& other) BOOST_NOEXCEPT :
+            thread_data(static_cast< data&& >(other.thread_data))
+        {}
+
+        id& operator=(id&& other) BOOST_NOEXCEPT
+        {
+            thread_data = static_cast< data&& >(other.thread_data);
+#if defined(BOOST_THREAD_PROVIDES_BASIC_THREAD_ID)
+            other.thread_data = 0;
+#endif
+            return *this;
+        }
+#endif // !defined(BOOST_NO_CXX11_RVALUE_REFERENCES)
+
```

Comment:
Can't we just remove the user defined copy constructor instead (it seems to be not doing anything special)

Reply:
The move assignment does have to do something special (assign zero to the move source), which means other operators/constructors are also needed.

Which makes me notice that the move constructor should set the move source to zero. I'll update the PR.